RFD: `tctl sso configure` command.

Tener opened the pull request:
This RFD is "implement auth connector configurator" part of #9270.

Tener committed:
First draft of 'sso configure' RFD

r0mant commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+The command will never modify existing Teleport configuration, this will be done by the user by further invocation
+of `tctl create` command or by other means.
+
+The command may handle user secrets. The implementation must ensure these are silently written anywhere (e.g. we must
```
Did you mean to say "these are **not** silently written anywhere"?

Tener replied:
Yep, good catch.

Tener commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+|          |            | Ask for `--kind` to use.                                  | Exit with error: missing `--kind`. |
+| `github` |            | Ask for `client_id` and `client_secret`.                  |                                    |
+| `oidc`   |            | Ask for `--preset`.                                       | Default to `--preset=generic`.     |
+| `oidc`   | `generic`  | Generic OIDC template.                                    |                                    |
```
I'll simplify and drop --kind entirely, so `--kind=oidc --preset=generic` will now be simply `--preset=oidc`.

Tener commented on rfd/0049-sso-configure-command.md:
````diff
@@ -0,0 +1,146 @@
+chooses template suited for particular vendor:
+
+```bash
+$ tctl sso configure --kind={oidc,saml,github} --preset={google,okta,adfs,generic} --interactive --template-specific-flags=...
````
Sounds good.

Tener commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+## Implementation details
+
+Each combination of `--kind` and `--provider` will be mapped to single template based
+on [text/template](https://pkg.go.dev/text/template) package. The extension points will be handled by calls to functions
```
Sure, this makes things simpler if less flexible (we cannot output ill-formed YAML with those methods, so there is no way of adding comments and similar stuff). But this may be a good thing after all: the implementation will be simpler and more testable.

Tener commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+
+## Proxy URL discovery
+
+The command will make an effort to discover the correct proxy URL. Failing that, an appropriate default along with
```
Ok.

Tener commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+
+## Interactive and batch modes
+
+We support two modes: interactive and batch (non-interactive). In interactive mode we may provide the user with
```
I rephrased this a bit.

Tener commented on rfd/0049-sso-configure-command.md:
```diff
@@ -0,0 +1,146 @@
+
+We support two modes: interactive and batch (non-interactive). In interactive mode we may provide the user with
+choices (e.g. preset) or ask them to provide details (secrets, identifiers etc.) In batch mode we never ask but use
+defaults (when possible) or insert comments asking for the data to be filled manually afterwards. Some
```
Thanks, this was a leftover from a deleted sentence.

Tener committed:
Changes from review.

Tener committed:
Merge branch 'master' into tener/rfd-sso-configure

Tener committed:
New RFD number.

Tener committed:
Update the RFD contents.

- The "interactive" mode is discarded.
- The SAML section describes flags and examples.
- Rephrased, updated other sections.

Tener commented:
@xinding33 @r0mant @smallinsky

Please take a look at this updated version.

r0mant commented on rfd/0062-tctl-sso-configure-command.md:
````diff
@@ -0,0 +1,205 @@
+- overrides for specific IdP providers
+
+```
+--signing_key_file         A file with request signing key. Must be used together with --signing_cert_file.
+--signing_cert_file        A file with request certificate. Must be used together with --signing_key_file.
+--assertion_key_file       A file with key used for securing SAML assertions. Must be used together with --assertion_cert_file.
+--assertion_cert_file      A file with cert used for securing SAML assertions. Must be used together with --assertion_key_file.
````
I'm pretty sure that our CLI flags use dashes instead of underscores so I'd rename all these new flags to match. I left the same comment in your implementation PR.

Tener replied:
Thanks, done.

r0mant reviewed (approved):
lgtm with the comment about flags.

r0mant commented:
@xinding33 Can you give this another quick look from the UX perspective and approve if it looks good so we can merge it? This basically describes the same `tctl sso configure` command we saw on a demo a couple weeks ago.

r0mant commented on rfd/0062-tctl-sso-configure-command.md:
```diff
@@ -0,0 +1,205 @@
+state: draft
+---
+
+# RFD 62 - `tctl sso configure` command
```
Sorry but I stole number 62 in https://github.com/gravitational/teleport/blob/master/rfd/0062-tsh-proxy-template.md so don't forget to update this :)

Tener committed:
Merge branch 'master' into tener/rfd-sso-configure

Tener committed:
Deallocate RFD number.

Tener committed:
Flags: underscores to dashes.

ben-mase commented on rfd/00XX-tctl-sso-configure-command.md:
```diff
@@ -0,0 +1,205 @@
+-p, --preset                   Preset. One of: [okta onelogin ad adfs]
+-n, --name                     Connector name. Required, unless implied from preset.
+-e, --entity-descriptor        Set the Entity Descriptor. Valid values: file, URL, XML content. Supplies configuration parameters as single XML instead of individual elements.
+-a, --attributes-to-roles      Sets attribute-to-role mapping in the form 'attr_name,attr_value,role1,role2,...'. Repeatable.
```
Looking at the examples, I find this syntax a bit hard to parse without going back and referencing the docs. I know it's tricky because the user is having to specify three or more things at once. There might not be a better solution, but I found myself wondering about alternative syntaxes like `-a group[developer]=access`

Tener replied:
Unfortunately, both `[` and `]` are special characters in many shells which makes them poor choices. But I'm happy to consider alternative syntaxes that do not have this problem.

Tener committed:
Merge branch 'master' into tener/rfd-sso-configure

Tener committed:
Allocate RFD number: 70.
